### PR TITLE
adding sdkClient for task manager to create task + updating snapshot + making master key id unique for multi_tenancy

### PR DESCRIPTION
**Pull request by dhrubo-os**
### Description
[
1. Adding sdkClient for task manager to create task
2. Updating snapshot to 2.16
3. making master key id unique for multi-tenancy
]

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java [hunk 1]
```diff
@@ -6,11 +6,12 @@
 package org.opensearch.ml.engine.encryptor;
 
 import static java.util.concurrent.TimeUnit.SECONDS;
-import static org.opensearch.ml.common.CommonValue.MASTER_KEY;
-import static org.opensearch.ml.common.CommonValue.ML_CONFIG_INDEX;
+import static org.opensearch.ml.common.CommonValue.*;
```
Comment (dhrubo-os) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java:9]: Will revert this change in the next commit.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java [hunk 2]
```diff
@@ -247,16 +251,46 @@ private void initializeNewMasterKey(
...
+    private String hashString(String input) {
+        try {
+            // Create a MessageDigest instance for SHA-256
+            MessageDigest digest = MessageDigest.getInstance("SHA-256");
+
+            // Perform the hashing and get the byte array
+            byte[] hashBytes = digest.digest(input.getBytes());
+
+            // Convert the byte array to a Base64 encoded string
+            return Base64.getEncoder().encodeToString(hashBytes);
```
Comment (dbwiddis) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java:287]: This creates slash characters in the encoding, which would break requests to a remote cluster going over the REST path.

File: plugin/src/main/java/org/opensearch/ml/action/model_group/TransportUpdateModelGroupAction.java [hunk 3]
```diff
@@ -206,29 +206,34 @@ private void updateModelGroup(
...
-                    updateModelGroup(modelGroupId, source, listener);
-                }
-            }, e -> {
-                log.error("Failed to search model group index", e);
-                listener.onFailure(e);
-            }));
+            mlModelGroupManager
+                .validateUniqueModelGroupName(
+                    updateModelGroupInput.getName(),
+                    updateModelGroupInput.getTenantId(),
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/action/model_group/TransportUpdateModelGroupAction.java:212]: Heh, I have this exact change on my local, looks like we're duplicating effort here.

File: plugin/src/test/java/org/opensearch/ml/action/model_group/TransportUpdateModelGroupActionTests.java [hunk 4]
```diff
@@ -179,10 +179,10 @@ public void setup() throws IOException {
 
         SearchResponse searchResponse = createModelGroupSearchResponse(0);
         doAnswer(invocation -> {
-            ActionListener<SearchResponse> listener = invocation.getArgument(1);
+            ActionListener<SearchResponse> listener = invocation.getArgument(2);
             listener.onResponse(searchResponse);
             return null;
-        }).when(mlModelGroupManager).validateUniqueModelGroupName(any(), any());
+        }).when(mlModelGroupManager).validateUniqueModelGroupName(any(), any(), any());
```
Comment (dbwiddis) [inline plugin/src/test/java/org/opensearch/ml/action/model_group/TransportUpdateModelGroupActionTests.java:185]: We'll probably want to split this out into two tests, with tenantID and with null, for the APIs that use tenant ID.

I'm already tracking that on my changes, so this PR is fine.

**Review by dbwiddis — COMMENTED** [review]
LGTM with the comments noted.

**Review by dbwiddis — APPROVED** [review]
LGTM, with the comment remaining about needing to add some tests for validating unique name both with and without model ID.  I am already tracking that in my integ test work.
